src/workerd/jsg/async-context.c++

```diff
@@ -0,0 +1,254 @@
+    auto thisArgHandle = v8::Private::ForApi(isolate, v8StrIntern(isolate, "thisArg"));
+
+    // We do not use the normal unwrapOpaque here since that would consume the wrapped
+    // value, and we need to be able to unwrap multiple times.
+    auto& frame = KJ_ASSERT_NONNULL(tryUnwrapFrame(isolate,
+        check(fn->GetPrivate(context, handle))));
+
+    v8::Local<v8::Value> thisArg = context->Global();
+    if (fn->HasPrivate(context, thisArgHandle).FromJust()) {
+      thisArg = check(fn->GetPrivate(context, thisArgHandle));
+    }
+
+    AsyncContextFrame::Scope scope(jsg::Lock::from(isolate), frame);
+
+    kj::Vector<v8::Local<v8::Value>> argv(args.Length());
+    for (int n = 0; n < args.Length(); n++) {
+      argv.add(args[n]);
+    }
+
+    v8::Local<v8::Value> result;
+    if (fn->Call(context, thisArg, args.Length(), argv.begin()).ToLocal(&result)) {
+      args.GetReturnValue().Set(result);
+    }
+  }, fn));
+}
+
+v8::Local<v8::Promise> AsyncContextFrame::wrap(
+    Lock& js,
+    v8::Local<v8::Promise> promise,
+    kj::Maybe<AsyncContextFrame&> maybeFrame) {
+  auto handle = v8::Private::ForApi(js.v8Isolate, v8StrIntern(js.v8Isolate, "asyncResource"));
+  auto context = js.v8Isolate->GetCurrentContext();
+  // If the promise has already been wrapped, do nothing else and just return the promise.
+  if (!check(promise->HasPrivate(context, handle))) {
+    // Otherwise, we have to create an opaque wrapper holding a ref to the current frame
+    // because we do not have the option of using an internal field with promises.
+    auto frame = kj::addRef(AsyncContextFrame::current(js));
+    KJ_ASSERT(check(promise->SetPrivate(context, handle, wrapOpaque(context, kj::mv(frame)))));
+  }
+  return promise;
+}
+
+kj::Maybe<Value&> AsyncContextFrame::get(StorageKey& key) {
+  KJ_ASSERT(!key.isDead());
+  storage.eraseAll([](const auto& entry) { return entry.key->isDead(); });
+  return storage.find(key).map([](auto& entry) -> Value& { return entry.value; });
+}
+
+AsyncContextFrame::Scope::Scope(Lock& js, AsyncContextFrame& resource)
+    : Scope(js.v8Isolate, resource) {}
+
+AsyncContextFrame::Scope::Scope(v8::Isolate* isolate, AsyncContextFrame& resource)
+    : isolate(IsolateBase::from(isolate)) {
+  this->isolate.pushAsyncFrame(resource);
+}
+
+AsyncContextFrame::Scope::~Scope() noexcept(false) {
+  isolate.popAsyncFrame();
+}
+
+AsyncContextFrame::StorageScope::StorageScope(
+    Lock& js,
+    StorageKey& key,
+    Value store)
+    : frame(AsyncContextFrame::create(js, nullptr, StorageEntry {
+        .key = kj::addRef(key),
+        .value = kj::mv(store)
+      })),
+      scope(js, *frame) {}
+
+void IsolateBase::pushAsyncFrame(AsyncContextFrame& next) {
+  asyncFrameStack.push_front(&next);
+}
+
+void IsolateBase::popAsyncFrame() {
+  asyncFrameStack.pop_front();
+  KJ_DASSERT(!asyncFrameStack.empty(), "the async context frame stack was corrupted");
+}
+
+void IsolateBase::promiseHook(v8::PromiseHookType type,
```

Comment:
Since this is a callback from V8, it needs to catch exceptions, especially `JsExceptionThrown`, otherwise exceptions will crash the server.